[7.x] SoftDeletable Many-To-Many Relationships

skaiblinger opened the pull request:
This PR tries to implement the full support of soft-deletes in many-to-many-relationships.

Related Issues: #725, #2733, #28407, and some questions about that feature on SO or laracasts.
Some background information was already gatered in https://github.com/laravel/ideas/issues/1908

**Additions** 
- New interface SoftDeleteable
Used to detect if the pivot model is soft deletable even if the SoftDeletes trait was not used (custom implementations)
- Ability to apply scopes and call additional methods on the pivot-query
This allows for calls like `$group->members()->pivotWithTrashed()->get()` or really any other method-call regarding the pivot query

**Breaking changes**

Since soft deletes were not supported in belongsToMany and morphToMany relationships, nothing should be breaking. But there is the possibility that some users have models which use the SoftDeletes trait or implemented something for themselves. 
Since this is my first contribution i thought pointing to 7.x would be on the safer side.

**Todo**
- [x] Tests - All existing and new database tests are passing. 
- [ ] Documentation - Remove the warning about soft deletes when using pivots and mention the SoftDeletable-Interface

skaiblinger committed:
Add ability to apply scopes on pivot table when using Pivot model class

- adds BelongsToMany::pivotModel and BelongsToMany::pivotQuery when
BelongsToMany::using() is called
- BelongsToMany::performJoin() uses the pivotQuery and joins it as
a subquery with the original tablename as alias
-- For the ability to toggle scopes during "configuration phase" of
the relation, the call to BelongsToMany::performJoin() had to be
moved from the constructors call chain, to the actual point where the
query is executed in BelongsToMany::get()
- BelongsToMany::__call() now treats calls to pivot* as calls which
should be forwarded to the pivot query. So things like eager loading
the relationship but configuring the SoftDeletes scope will work

skaiblinger committed:
Merge remote-tracking branch 'remotes/upstream/master'

skaiblinger committed:
Merge remote-tracking branch 'remotes/upstream/master'

skaiblinger committed:
Merge remote-tracking branch 'remotes/upstream/master'

# Conflicts:
#	src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php

skaiblinger committed:
Implemented restoring of soft deleted pivots in attach

 - NEW: Interface SoftDeletable
 -- Used by InteractsWithPivotTable::attachUsingCustomClass() to check
 if the pivot model is soft deletable
 - CHANGED: InteractsWithPivotTable::attachUsingCustomClass() checks
 if the pivot model implements the interface SoftDeletable and tries to
 restore deleted pivots. All not found records will be inserted as
 before. All found but not trashed pivots will be ignored for now.

skaiblinger committed:
Working on sync with trashed pivots:

 - InteractsWithPivotTable::newPivotStatement() now uses the static
 query method of the custom class defined in
 BelongsToMany::using to get a query with applied global scopes
 -- InteractsWithPivotTable::getCurrentlyAttachedPivots() uses that
 query to only load the "active" records
 -- InteractsWithPivotTable::attach() uses
  InteractsWithPivotTable::newPivotQuery() to find all trashed pivots
  which should be restored
 -- BelongsToMany::using() now makes use of
 InteractsWithPivotTable::newPivotStatement()
 - Replaced BelongsToMany::pivotModel with BelongsToMany::pivotTable
 as we don't need anything but the table from the model.
 - InteractsWithPivotTable::attach() now updates and restores trashed
 pivots. Already attached and not trashed pivots will now again be
 inserted and will fail on unique-constraint violations, which is the
 wanted behaviour for attach

skaiblinger committed:
Fixed missing return in InteractsWithPivotTable::newPivotStatement()

skaiblinger committed:
Merge remote-tracking branch 'upstream/master'

skaiblinger committed:
Calling performJoin() in addConstraints() again

 - Many methods, apart from attach, detach and sync depend on the joins
 - Instead recreate the query in using() when using() was called after
 the constructor already ran. For example when the relation is called
 like "return new belongsToMany('App\Related')->using('...')"

skaiblinger committed:
Delayed creation of the query-joins and the application of scopes again

 - Implemented BelongsToMany::configureQuery() to do this tasks
 -- This method can be called when the query is fully configured and is
 about to be executed
 -- All methods which excute the query like chunk, chunkById, cursor,
 ... now call this method first
 - Removed expected calls to join from the tests which really don't need
 the joins to be in place.
 -

skaiblinger committed:
Implemented tests for the new features

skaiblinger committed:
style-ci fixes

skaiblinger committed:
applied style-ci diff

driesvints commented:
@staudenmeir any chance you can give this a review? Thanks :)

GrahamCampbell commented on src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php:
```diff
@@ -184,13 +198,29 @@ protected function resolveTableName($table)
+     *
+     * If this steps are executed before everything is configured the resulting join may still contain where clauses
+     * of scopes which are disabled later in the process.
+     */
```
Missing return annotation.

GrahamCampbell commented on src/Illuminate/Database/Eloquent/Relations/Concerns/InteractsWithPivotTable.php:
```diff
@@ -273,10 +274,26 @@ public function attach($id, array $attributes = [], $touch = true)
     {
+        $idsToAttach = $this->parseIds($id);
+
+        //check if pivot model uses soft deleting
```
missing space after `//`

skaiblinger commented:
Since it's my first real deep dive into the framework source, the chance of me missing some tests / use-cases is quiet substantial.

So i will gladly take all reviews, comments, testers and give my best to make this work.

skaiblinger committed:
Fixes to comments / code-style

staudenmeir commented:
@driesvints I can review the PR, but only if @taylorotwell actually plans to merge it.

driesvints commented:
@staudenmeir fair enough. I'll ping him once he's online.

taylorotwell commented on src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php:
```diff
@@ -324,6 +360,15 @@ public function getPivotClass()
+        if ($this->query) {
+            $this->query = $this->related->newModelQuery();
+            $this->addConstraints();
+        }
```
This feels pretty hacky and will lose any custom `where` that was added in the relationship method.

skaiblinger replied:
Removing the "late" using-call would have been much cleaner but would have changed the interface.

taylorotwell replied:
Do you have any comments regarding breaking `wheres` that are added during the relationship definition?

taylorotwell replied:
Or really, any other type of query element... orderBys, etc.

skaiblinger replied:
Unless i miss something, all the other wheres, orderBys and such should not be removed if the the pivot is configured during the constructor. 

It should even work if it is used like `return $this->belongsToMany()->using()->where()->orderBy()`

skaiblinger replied:
I removed this part, since it is no longer necessary. 
I already moved the creation of the join / joinSub out of the constructor call-chain and delayed that part to the point where the query will actually be executed. 
All tests are still passing, so i'm confident i didn't break the new features again ;)

taylorotwell commented:
There are some hacky things in here I'm not comfortable taking on the responsibility to maintain from this point forward. You can already have your own version of soft deletable many to many relationships by using a custom Pivot class and `hasManyThrough`.

skaiblinger commented:
> 
> 
> There are some hacky things in here I'm not comfortable taking on the responsibility to maintain from this point forward. 

If there is a chance to get the hacky things better implemented i would love to fix these up.
Could you give me some hints which parts cause the troubles?

> 
> 
> You can already have your own version of soft deletable many to many relationships by using a custom Pivot class and `hasManyThrough`.

A custom pivot and `hasManyThrough`? I never found a satisfing soluting for all the relations in my project. Do you have a link handy?

taylorotwell commented:
If you're not using custom pivot models what are your options? Could we chain a method ->withSoftDeletes() onto the relationship definition like another PR attempted to do?

skaiblinger commented:
Hmm ... at first glance, merging the features of #31210 should not be to difficult.

But i tried not to touch the "default" behavior without custom pivot models. So i will have to take a closer look.

skaiblinger committed:
Merge remote-tracking branch 'remotes/upstream/master'

skaiblinger committed:
Ditch query recreation when using is called outside of the constructor

Since the joins will be created when the results are requested and not
already in the constructor chain, recreating the query and potentially
loosing already added wheres, orderBys etc, is not necessary

taylorotwell commented:
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If possible, please consider releasing your code as a package so that the community can still take advantage of your contributions! 

If you feel absolutely certain that this code corrects a bug in the framework, please "**@**" mention me in a follow-up comment with further explanation so that GitHub will send me a notification of your response.

alex-zukitek commented:
@skaiblinger , Any updating or your package?

skaiblinger commented:
@alex-zukitek No sorry. The PR #31219 solved most problems for most users. If there is a need for the more complex scenarios like the pivot scopes or making the soft-deletable attribute customizable i could give this another try.
